Bowling/BowlingGame.cs

```diff
@@ -0,0 +1,150 @@
+        private int bonus = -1;
+        public int NumberOfPlayers { get; set; }
+        public int NumberOfFrames { get; set; }
+        public BowlingGame(int numberOfFrames, int numberOfPlayers)
+        {
+            if (numberOfPlayers < 1)
+            {
+                throw new NotEnoughPlayersException(Constants.numberOfPlayersMessage);
+            }
+            else
+            {
+                NumberOfPlayers = numberOfPlayers;
+            }
+
+            if (numberOfFrames < 1 || numberOfFrames > 10)
+            {
+                throw new NumberOfFramesOutOfRangeException(Constants.numberOfFramesMessage);
+            }
+            else
+            {
+                NumberOfFrames = numberOfFrames;
+                InitializeFrames(NumberOfFrames);
+            }
+        }
+
+        private void InitializeFrames(int numberOfFrames)
+        {
+            for (var frameIndex = 0; frameIndex < numberOfFrames; frameIndex++)
+                framesScoreBoard.Add(new Frame());
+        }
+
+        public void AddPlayer(Player player)
+        {
+            if (NumberOfPlayers == players.Count)
+            {
+                throw new AddedMorePlayersThanDeclaredException(Constants.declaredAnotherNumberOfPlayers);
+            }
+            else
+            {
+                players.Add(player);
+            }
+        }
+
+        public void ShowPlayers()
+        {
+            foreach (Player player in players)
+            {
+                Console.WriteLine(player.Name);
+            }
+        }
+
+        public void Roll(int pins)
+        {
+            if (framesScoreBoard[CurrentFrameIndex].CanInsert())
+            {
+                framesScoreBoard[CurrentFrameIndex].Insert(pins);
+            }
+            else
+            {
+                if (CurrentFrameIndex == NumberOfFrames - 1)
+                {
+                    RollForLastBonus(pins);
+                }
+                else
+                {
+                    RollForNextFrame(pins);
+                }
+            }
+        }
+
+        private void RollForLastBonus(int pins)
+        {
+            if (CanRollForLastBonus() == true)
+            {
+                bonus = pins;
+            }
+            else
+            {
+                throw new UnfairBonusRollException(Constants.unfairBonusMessage);
+            }
+        }
+
+        private void RollForNextFrame(int pins)
+        {
+            CurrentFrameIndex += 1;
+            framesScoreBoard[CurrentFrameIndex].Insert(pins);
+        }
+
+        public void DisplayScore()
+        {
+            var numFramesExceptLast = NumberOfFrames - 1;
+            for (var frameIndex = 0; frameIndex < numFramesExceptLast; frameIndex++)
+            {
+                var frameShots = framesScoreBoard[frameIndex];
+                Console.WriteLine(String.Format(Constants.scoreDisplayFormat, frameShots.FirstShot, frameShots.SecondShot, CalculateFrameScore(frameIndex)));
```

Comment:
Now when you want to display the score frame you calculate it each time. 
Do you think we can keep the values in a data structure to have access faster to the score frame ?
This maybe it will help to display the score now that you make the game to be multiplayer.